core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -174,13 +175,18 @@ public void clusterChanged(final ClusterChangedEvent event) {
 
             cleanFailedShards(event);
 
+            // cleaning up indices that are completely deleted so we won't need to worry about them
+            // when checking for shards
             applyDeletedIndices(event);
+            applyDeletedShards(event);
+            // call after deleted shards so indices with no shards will be cleaned
+            applyCleanedIndices(event);
+            // make sure that newly created shards use the latest meta data
+            applyIndexMetaData(event);
             applyNewIndices(event);
+            // apply mappings also updates new indices. TODO: make new indices good to begin with
             applyMappings(event);
```

Comment:
I'm with fine with doing that after the review is done. The advantage for rearranging the methods is that in future PRs it is easier to see how changes to two or more of these methods affect the overall behavior as the changes are then displayed in Github UI in the order in which the code is executed.